code/controllers/subsystem/fluids.dm

```diff
@@ -0,0 +1,266 @@
+		currently_spreading = spread_carousel[spread_bucket_index]
+		spread_carousel[spread_bucket_index] = list() // Reset the bucket so we don't process an _entire station's worth of foam_ spreading every 2 ticks when the foam flood event happens.
+		resumed_spreading = TRUE
+
+	delta_time = spread_wait / (1 SECONDS)
+	currentrun = currently_spreading
+	while(currentrun.len)
+		var/obj/effect/particle_effect/fluid/to_spread = currentrun[currentrun.len]
+		currentrun.len--
+
+		if(!QDELETED(to_spread))
+			to_spread.spread(delta_time)
+			to_spread.spread_bucket = null
+
+		if (MC_TICK_CHECK)
+			break
+
+	if(!currentrun.len)
+		resumed_spreading = FALSE
+
+	MC_SPLIT_TICK // Start processing fluid effects:
+	if(!resumed_effect_processing)
+		effect_bucket_index = (effect_bucket_index % num_effect_buckets) + 1
+		var/list/tmp_list = effect_carousel[effect_bucket_index]
+		currently_processing = tmp_list.Copy()
+		resumed_effect_processing = TRUE
+
+	delta_time = effect_wait / (1 SECONDS)
+	cached_bucket_index = effect_bucket_index
+	currentrun = currently_processing
+	while(currentrun.len)
+		var/obj/effect/particle_effect/fluid/to_process = currentrun[currentrun.len]
+		currentrun.len--
+
+		if (QDELETED(to_process) || to_process.process(delta_time) == PROCESS_KILL)
+			effect_carousel[cached_bucket_index] -= to_process
+			to_process.effect_bucket = null
+			to_process.datum_flags &= ~DF_ISPROCESSING
+
+		if (MC_TICK_CHECK)
+			break
+
+	if(!currentrun.len)
+		resumed_effect_processing = FALSE
+
+
+/**
+ * Queues a fluid node to spread later after one full carousel rotation.
+ *
+ * Arguments:
+ * - [node][/obj/effect/particle_effect/fluid]: The node to queue to spread.
+ */
+/datum/controller/subsystem/fluids/proc/queue_spread(obj/effect/particle_effect/fluid/node)
+	if (node.spread_bucket)
+		return
+
+	spread_carousel[spread_bucket_index] += node
+	node.spread_bucket = spread_bucket_index
+
+/**
+ * Cancels a queued spread of a fluid node.
+ *
+ * Arguments:
+ * - [node][/obj/effect/particle_effect/fluid]: The node to cancel the spread of.
+ */
+/datum/controller/subsystem/fluids/proc/cancel_spread(obj/effect/particle_effect/fluid/node)
+	if(!node.spread_bucket)
+		return
+
+	var/bucket_index = node.spread_bucket
+	spread_carousel[bucket_index] -= node
+	if (bucket_index == spread_bucket_index)
+		currently_spreading -= node
+
+	node.spread_bucket = null
+
+/**
+ * Starts processing the effects of a fluid node.
+ *
+ * The fluid node will next process after one full bucket rotation.
+ *
+ * Arguments:
+ * - [node][/obj/effect/particle_effect/fluid]: The node to start processing.
+ */
+/datum/controller/subsystem/fluids/proc/start_processing(obj/effect/particle_effect/fluid/node)
+	if (node.datum_flags & DF_ISPROCESSING || node.effect_bucket)
+		return
+
+	var/bucket_index = rand(1, num_effect_buckets) // Edit this value to make all fluids process effects (at the same time|offset by when they started processing| -> offset by a random amount <- )
```

Comment:
mind moving this comment up a line? makes it easier to read

Reply:
Sure. Done.